[GraphQL] Compare products

Usik2203 opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Based on this schema [compare-list.graphqls](https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/catalog/compare-list.graphqls) we should provide functionality for `list_id` parameter in schema
New schema https://github.com/magento/architecture/pull/430/files
Purpose `list_id` is next:
<pre>
The list_id becomes necessary since we want to move out the state from 
the app we have some freedom to build define a new schema GraphQL schema. 
The same as it was before for card, we need somehow to identify the list in the operations. 
Compare list may belong to a registered customer or a guest for both cases set of operation will be the same,
 technically there is no sense to associate compare list with a client unless you want to save the list for future usage. 
This is why we need to expose a new identifier list_id.
This new Id requires a new table that will have relation one to many (comparable items) and 
will represent the list as an entity.
</pre> 
This PR  adds `list_id` parameter during comparing products and add **CompareListGraphQl** module


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#28522>

### Manual testing scenario
**Create Compare List**
<pre>
mutation{
  createCompareList (input: { products: ["123", "456"] }) #products optional {
      uid
        items {
	uid
      	product {
              id
              sku
            }
      	attributes {
          code
          value
        }
      }
    attributes {
      code
      label
    }
  }
}
</pre>


**Getting compared list for Customer**

<pre>
{
  customer {
    firstname
    lastname
    
    compare_list {
      uid 
          items {
	uid
      	product {
              id
              sku
            }
      	attributes {
          code
          value
        }
      }
    attributes {
      code
      label
    }
    }
  }
}
</pre>
**Get compareList by ID**
<pre>
{
  compareList(uid: hash) {
    uid
           items {
	uid
      	product {
              id
              sku
            }
      	attributes {
          code
          value
        }
      }
    attributes {
      code
      label
    }
  }
}

</pre>

**Add Products to compare list**
<pre>
mutation {
  addProductsToCompareList(input: { uid: hash, products: [ 2049,  2048 ]}) {
    uid
        items {
	uid
      	product {
              id
              sku
            }
      	attributes {
          code
          value
        }
      }
    attributes {
      code
      label
    }
  }
}
</pre>
**Remove items from compare list**
<pre>
mutation {
   removeProductsFromCompareList(input: {uid: "hash" , products: [2047, 2048]}) {
      uid
         items {
	uid
      	product {
              id
              sku
            }
      	attributes {
          code
          value
        }
      }
    attributes {
      code
      label
    }
  }
}

</pre>

**Assign compare list to customer**
<pre>
mutation {
  assignCompareListToCustomer(uid: hash)
}
</pre>

**Delete compare list**
<pre>
mutation {
  deleteCompareList(uid:"zo8OIjXQR0r04lx8GEbg9Ivy0HabwFnS") {
    result
  }
}
</pre>

Usik2203 committed:
Add DB schema

Usik2203 committed:
add list_id during comparing products

Usik2203 commented:
@magento run all tests

Usik2203 committed:
set customer for compare list during login

Usik2203 commented:
Hi @rogyar 
Could I kindly ask you to review this PR if you have slot of free time?
Perhaps you will advise me something before I will start to do GraphQl coverage based on this functional.
Thanks a lot!

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/AddToList.php:
```diff
@@ -0,0 +1,154 @@
+use Magento\Catalog\Model\CompareList;
+use Magento\Catalog\Model\CompareListFactory;
+use Magento\Catalog\Model\ResourceModel\CompareList as CompareListResource;
+use Magento\Customer\Model\Session;
```
Instead of bringing the session to this service, can we pass visitor id or customer id on following stage to this class somehow? 

https://github.com/magento/magento2/blob/da430478c745293769787f6e2309575ab14ba8ea/app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php#L107

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/AddToList.php:
```diff
@@ -0,0 +1,154 @@
+use Magento\Customer\Model\Session;
+use Magento\Customer\Model\Visitor;
+
+class AddToList
```
The name of the class says `AddToList` but the `execute` method of this class responsible for returning the list ID. I assume we need to change the class name

Usik2203 replied:
**Done** renamed to `CompareList`

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/AddToList.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * Set customer from visitor
+     */
+    public function setCustomerFromVisitor()
```
I would recommend extracting this logic to a separate class

Usik2203 replied:
**Done** moved this method to `\Magento\Catalog\Model\ResourceModel\Product\Compare\Item`

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/AddToList.php:
```diff
@@ -0,0 +1,154 @@
+     */
+    public function setCustomerFromVisitor()
+    {
+        $customerId = $this->customerSession->getCustomerId();
```
I would rather try to pass this information from the invocation place. Maybe we may get this data in the similar way

https://github.com/magento/magento2/blob/2fe99e93914d4b2c2a548c97ef34dca391f64ddb/app/code/Magento/Catalog/Model/ResourceModel/Product/Compare/Item.php#L216

So we don't inject the session over and over again to the service classes.

Usik2203 replied:
**Done**

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 committed:
changes based on review

Usik2203 commented:
@magento run all tests

Usik2203 committed:
started working on compared list graphql modules

Usik2203 commented:
@magento run all tests

Usik2203 committed:
added resolver for getting compare list by id

Usik2203 committed:
add remove and assign functionality

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Resolver/AddItemsToCompareList.php:
```diff
@@ -0,0 +1,155 @@
+
+namespace Magento\CompareListGraphQl\Model\Resolver;
+
+use Magento\Catalog\Model\CompareList as ModelCompareList;
```
I would recommend decoupling this resolver into smaller chunks. We may make the resolver 'thin' by moving all the business logic to a separate class(es). In that way we will reduce the number of dependencies in this resolver

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Resolver/AssignCompareListToCustomer.php:
```diff
@@ -0,0 +1,160 @@
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+
+class AssignCompareListToCustomer implements ResolverInterface
```
The same advice is applicable for this resolver and other resolvers where you see the number of dependencies more than 13. We may think about moving all the business logic away from resolver

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Service/CompareListService.php:
```diff
@@ -0,0 +1,272 @@
+     * @param CompareList                   $compareList
+     * @param ListCompare                   $listCompare
+     */
+    public function __construct(
```
This service contains more dependencies than recommended for a single class as well. Also, I see that there are two entry points (public functions). In order to follow the Separation of Concerns principle I would strongly recommend splitting this service into (at least) two services with one entry point per service correspondingly.

Usik2203 replied:
**Done**

Usik2203 committed:
separated some logic

Usik2203 committed:
have decoupled some resolvers

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/CompareList.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return int
+     */
+    public function getListId(Item $item)
```
As far as I see, this class has the only one public method for retrieving the list ID. I would suggest renaming the class to `GetItemListId()`.
Also, this class has a new compare list creation as side-effect. I would not recommend hiding it in the service for retrieving list id. This particular class should return either ID or null/exception in case if the list does not exist. 
Then, the parent class responsible for retrieving the list ID should create a new list if there's no exiting list yet. Something like:

```php
$listId = GetItemListId::execute($item);
if ($listId === NULL) {
    $listId = // Create new list 
}

rogyar replied:
Hi @Usik2203. Please, take a look at my suggestion above

Usik2203 replied:
File was removed from PR

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Service/AddToCompareListService.php:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Service add product to compare list
+ */
+class AddToCompareListService
```
You may use `AddToCompareList` as a name, without suffixing with the "...Service". Also, I would recommend renaming the entry point (public function) to `execute` in this way we will have more or less consistent naming for the method invocation in other classes

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Service/Collection/ComparableItems.php:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Get collection with comparable items
+ */
+class ComparableItems
```
Please, make sure that the service name reflects its purpose. It's recommended to use a verb in the very beginning of the service like `GetComparableItemsCollection`. Here as well, you may use `execute` method as the entry point. 

This recommendation apply to all other services.

Thank you

Usik2203 replied:
**Done**

prabhuram93 commented on app/code/Magento/Catalog/etc/db_schema.xml:
```diff
@@ -573,6 +577,27 @@
+            <column name="list_id"/>
+        </index>
+        <index referenceId="CATALOG_COMPARE_LIST_VISITOR_ID_CUSTOMER_ID" indexType="btree">
+            <column name="visitor_id"/>
```
@Usik2203 Why do we need the visitor_id? How is it being generated for a GraphQl client?

Usik2203 replied:
Hi @prabhuram93 
`visitor_id` it's value come from table `customer_visitor` field `visitor_id` 
It need for detecting what compare list should remove or add to compare list for guest users

Scenario: https://drive.google.com/file/d/1jfEdoXa4tArU7z1rhSHgZX08Gxr4DzQL/view?usp=sharing

Usik2203 replied:
**UPD**
@prabhuram93 
I have got rid from `visitor_id` field in` DB`  as redundant after applying changes based on new schema
Thanks

prabhuram93 commented on app/code/Magento/CompareListGraphQl/Model/Resolver/AddItemsToCompareList.php:
```diff
@@ -0,0 +1,110 @@
+        array $value = null,
+        array $args = null
+    ) {
+        $listId = (int)$args['id'];
```
@Usik2203 When clients are using this resolver for the first time, how will they get the listId?

Usik2203 replied:
@prabhuram93 I was also confused about this case.  
I see that perhaps new mutation will be added for this purpose.
https://github.com/magento/architecture/pull/430/files#diff-184217e17b6d264d529a8363a0bae585R31

prabhuram93 commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,42 @@
+}
+
+type Mutation {
+    addItemsToCompareList(id: ID!, items: [ID!]): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddItemsToCompareList") @doc(description: "Add items to compare list")
```
Is this id being generated on the client side? How will guest users have access to this id?

Usik2203 replied:
@prabhuram93 
Unfortunately I didn't take this case into account
Could you advise please how this case should work  ?

nrkapoor commented:
@magento run all tests

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 committed:
Added changes to schema and add some class

Usik2203 committed:
set type list_id as varchar in DB

nrkapoor commented:
@magento run all tests

Usik2203 committed:
add changes to DB scheme

nrkapoor commented:
@magento run all tests

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/CompareList.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return int
+     */
+    private function getListIdByVisitorId($visitorId)
```
Please, use type hinting for parameters and return types. It applies to all similar places

Usik2203 replied:
File was removed from PR

rogyar commented on app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php:
```diff
@@ -77,6 +83,7 @@ public function __construct(
         \Magento\Catalog\Model\ResourceModel\Product\Compare\Item $catalogProductCompareItem,
         \Magento\Customer\Model\Session $customerSession,
         \Magento\Customer\Model\Visitor $customerVisitor,
+        \Magento\Catalog\Model\Product\Compare\CompareList $compareList,
```
Please, inject the `$compareList` at the very end of the parameters list with default `null` value. Then, init it via object manager in the same way as it's done here

https://github.com/magento/magento2/pull/29047/files#diff-e8d4535eb252d8b7fb3053dfdd7408cdba509e40d3adfc0bca25db38124f86f5R96

It's required for keeping backward compatibility

Usik2203 replied:
Changes were reverted from this file

rogyar commented on app/code/Magento/Catalog/Model/ResourceModel/CompareList.php:
```diff
@@ -0,0 +1,23 @@
+
+use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
+
+
```
Please, check the extra empty lines

Usik2203 replied:
File was removed

rogyar commented on app/code/Magento/Catalog/Model/ResourceModel/Product/Compare/Item.php:
```diff
@@ -242,4 +274,28 @@ public function clearItems($visitorId = null, $customerId = null)
+     *
+     * @return $this
+     */
+    private function setCustomerFromVisitor($item)
```
Please, use the type hinting

Usik2203 replied:
Changes were reverted for this file

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Resolver/AssignCompareListToCustomer.php:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        $compareListModel->setCustomerId($customer->getId());
+        $this->resourceCompareList->save($compareListModel);
```
Please, make sure that we handle all the exceptions here and in similar places.

Usik2203 replied:
**done**

rogyar commented:
Hi @Usik2203. After another round of review, I assume that this PR is still in progress. 
Please, let me know when it's ready for the review. Make sure that you use type hinting, exceptions handling on resolver level everywhere it's necessary. 

Thanks

Usik2203 committed:
Extract list id to list_id_mask table

Usik2203 committed:
reverted changes

Usik2203 committed:
create compare list

Usik2203 committed:
added resolver

Usik2203 committed:
rename classes

Usik2203 commented:
@magento run all tests

Usik2203 committed:
Refactoring resolvers

Usik2203 commented:
@magento run all tests

Usik2203 committed:
removing product from compare list and compare list

Usik2203 commented:
@magento run all tests

Usik2203 committed:
fix static issues

Usik2203 commented:
@magento run all tests

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

Usik2203 committed:
static issue fixes

Usik2203 committed:
changes in DB scheme

Usik2203 commented:
@magento run all tests

Usik2203 commented:
@magento run Integration Tests

prabhuram93 commented:
@Usik2203 Let us know when it is ready for reviewing. Thanks!

Usik2203 committed:
Minor changes in schema

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run Integration Tests

Usik2203 commented:
@magento run all tests

Usik2203 commented:
Hi @rogyar  , @prabhuram93 
This PR is ready for review 
I'm started working on test coverage for this one.
Thanks

Usik2203 commented:
Hi @prabhuram93  !
https://github.com/magento/architecture/pull/430/files
Could you clarify please logic of this mutation 
`assignCompareListToCustomer(uid: ID!): CompareList # Customer token needs to be passed`
During realisation this mutation i passed UID of compare list 
My logic were next :
After customer logged `assignCompareListToCustomer` will be call 
we will pass `UID of compare list` and customer id will be get from` $context `
Such way we will assign customer to current  compare list 

But if I pass Customer token  how i should get current compare list ?
Could you clarify this moment please ?
Thanks

prabhuram93 commented:
> Hi @prabhuram93 !
> https://github.com/magento/architecture/pull/430/files
> Could you clarify please logic of this mutation
> `assignCompareListToCustomer(uid: ID!): CompareList # Customer token needs to be passed`
> During realisation this mutation i passed UID of compare list
> My logic were next :
> After customer logged `assignCompareListToCustomer` will be call
> we will pass `UID of compare list` and customer id will be get from`$context`
> Such way we will assign customer to current compare list
> 
> But if I pass Customer token how i should get current compare list ?
> Could you clarify this moment please ?
> Thanks

Hey @Usik2203. Your logic is correct. The uid that needs to be passed is that of the compareList. I know the comment is vague `# Customer token needs to be passed`, but it was supposed to convey customer token needs to be passed in the headers. I will get it fixed. Thanks!

Usik2203 committed:
Add changes after schema updating

Usik2203 commented:
@magento run all tests

Usik2203 committed:
Add test coverage

Usik2203 commented:
@magento run all tests

Usik2203 commented:
@magento run Integration Tests

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

Usik2203 committed:
add new scenario to test coverage

Usik2203 commented:
@magento run all tests

rogyar commented on app/code/Magento/Catalog/Model/CompareListIdToMaskedListId.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        $listIdMask = $this->listIdMaskFactory->create();
+        $this->listIdMaskResource->load($listIdMask, $listId, 'list_id');
+        return $listIdMask->getMaskedId() ?? '';
```
I would suggest returning `null` instead of empty string. In that way you will make behaviour of this method more predictable

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/Catalog/Model/MaskedListIdToCompareListId.php:
```diff
@@ -0,0 +1,53 @@
+        $listIdMask = $this->listIdMaskFactory->create();
+        $this->listIdMaskResource->load($listIdMask, $maskedListId, 'masked_id');
+
+        return (int)$listIdMask->getListId();
```
Have you tested the scenario when compare list ID is missing in the DB for the provided masked value?

Usik2203 replied:
Hi @rogyar 
We will get `0` in this case 
<img width="1423" alt="Screenshot 2020-11-04 at 17 55 21" src="https://user-images.githubusercontent.com/13517181/98134077-09ff8d00-1ec7-11eb-80ac-36d8e5795b6e.png">

rogyar commented on app/code/Magento/Catalog/Model/ResourceModel/Product/Compare/Item/Collection.php:
```diff
@@ -156,6 +163,30 @@ public function setCustomerId($customerId)
+     *
+     * @return int
+     */
+    public function getListId()
```
Please, specify the return type

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/Catalog/Model/ResourceModel/Product/Compare/Item/Collection.php:
```diff
@@ -232,6 +267,26 @@ public function _addJoinToSelect()
+     *
+     * @return array
+     */
+    public function getProductsByListId(int $listId)
```
Please, specify the return type

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Resolver/AddProductsToCompareList.php:
```diff
@@ -0,0 +1,95 @@
+        array $args = null
+    ) {
+        $storeId = (int)$context->getExtensionAttributes()->getStore()->getStoreId();
+        if (!isset($args['input']['uid'])) {
```
This check will be performed automatically on the schema level. We should rather check if the value is not empty. Please, check other similar places as well

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Resolver/DeleteCompareList.php:
```diff
@@ -0,0 +1,99 @@
+            throw new GraphQlInputException(__('"uid" value must be specified'));
+        }
+
+        $listId = $this->maskedListIdToCompareListId->execute($args['uid']);
```
I would add an additional check for verifying that the provided list belongs to the current customer (in case if the compare list has customer ID). It will help to avoid cases when guest can remove a compare list of registered user.

rogyar replied:
The same suggestion is applicable for other mutations.

Usik2203 replied:
**Done** Additional check was added

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Service/CustomerService.php:
```diff
@@ -0,0 +1,158 @@
+/**
+ * Service class for customer
+ */
+class CustomerService
```
I would strongly suggest splitting this class into 3 different classes:
- GetListIdByCustomerId
- SetCustomerToCompareList
- ValidateCustomer

Usik2203 replied:
**Done**

rogyar commented on app/code/Magento/CompareListGraphQl/Model/Service/GetComparableItems.php:
```diff
@@ -0,0 +1,133 @@
+     * @return array
+     * @throws GraphQlInputException
+     */
+    public function getComparableItems(int $listId, ContextInterface $context)
```
Please, rename it to `execute()`. We already have a meaningful name of the class itself.

Usik2203 replied:
**Done**

rogyar reviewed (changes requested):
Hi @Usik2203. Awesome piece of work! Please, check my suggestions and take a look at the failing tests. 

Thank you!

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run Integration Tests

prabhuram93 commented:
@Usik2203 First of all, Amazing work. Very impressive 👏 
I have a couple of comments. @rogyar @cpartica can pitch in as well.

Based on the current implementation, the Luma ecosystem and the new GraphQl version are not integrated. But it can be done with a few small tweaks to the current implementation.

**Usecase  1:** 
When a new list is being created via GQL for a customer, who already has a comparison list in Luma, I would suggest updating those existing records to point to the new list.
<img width="449" alt="Screen Shot 2020-11-02 at 4 58 42 PM" src="https://user-images.githubusercontent.com/43612864/97928449-b0635b00-1d2c-11eb-98a6-05aacd2eaaf1.png">

On the above, records (5) belongs to customer 42 in Luma, 7 belongs to customer 42 in GQL implementation. 

TODO: So when creating a new list for a customer, check for existing records belonging to the customer, if found (can be more than one), update those records with the newly created list_id. On above example, 5 should have list_id, 7 should have customer_id.

**Usecase 2:**
When adding a product to the list in GQL, you are currently checking if the product already exists on the new implementation.

TODO: Nothing needs to be done for Luma change! if TODO on usecase 1 is performed this will work seamlessly.

**Usecase 3**
This is where it gets tricky. When **all** the list items for the customer are deleted in Luma, the entry in compare_list table needs to be removed(just a single row, can exist, still). Some logic needs to be added for this.
<img width="415" alt="Screen Shot 2020-11-02 at 5 21 30 PM" src="https://user-images.githubusercontent.com/43612864/97929915-e229f100-1d2f-11eb-887d-b1a817b3dc02.png">

If the above use cases are handled we can have an integrated system between Luma and the new GQL version.

@rogyar @cpartica, Please add your thoughts.

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run WebAPI Tests

Usik2203 committed:
Fixed failed test due to ProductInterface in schema

Usik2203 commented:
@magento run WebAPI Tests

Usik2203 commented:
@magento run Integration Tests

Usik2203 committed:
CR recommendations

Usik2203 commented:
@magento run all tests

prabhuram93 commented on app/code/Magento/Catalog/etc/db_schema.xml:
```diff
@@ -573,6 +577,40 @@
+            <column name="customer_id"/>
+        </index>
+    </table>
+    <table name="list_id_mask" resource="default" engine="innodb" comment="List ID and masked ID mapping">
```
@Usik2203 why do we need a separate table for this. A new column on catalog_compare_list should suffice right?

Usik2203 replied:
@prabhuram93  
I followed example of `createEmptyCart` mutation during creation `createCompareList` mutation

 It seems me that logic of `createEmptyCart` is very similar with `createCompareList` .

`createEmptyCart` creates and stores hash of quotes in` quote_id_mask` table

I thought that it's good example and created `list_id_mask` table for storing hash of compare lists

Thanks

prabhuram93 replied:
@Usik2203 It was performed that way in quote, because id masking was an additive feature to quote and "quote" table existed before that feature. Since we are introducing "catalog_compare_list" in this pr, it makes sense to add one more column for the mask, instead of having a new table.

Usik2203 replied:
I will do this change.
Thanks

Usik2203 replied:
**Done** Table was removed

Usik2203 committed:
Removed list id mask table

Usik2203 committed:
add additional check

Usik2203 commented:
@magento run all tests

Usik2203 committed:
luma compatibility add list id for customer compare list

Usik2203 commented:
@magento run all tests

Usik2203 committed:
luma compatibility remove compare list

Usik2203 commented:
> @Usik2203 First of all, Amazing work. Very impressive 👏
> I have a couple of comments. @rogyar @cpartica can pitch in as well.
> 
> Based on the current implementation, the Luma ecosystem and the new GraphQl version are not integrated. But it can be done with a few small tweaks to the current implementation.
> 
> **Usecase 1:**
> When a new list is being created via GQL for a customer, who already has a comparison list in Luma, I would suggest updating those existing records to point to the new list.
> <img alt="Screen Shot 2020-11-02 at 4 58 42 PM" width="449" src="https://user-images.githubusercontent.com/43612864/97928449-b0635b00-1d2c-11eb-98a6-05aacd2eaaf1.png">
> 
> On the above, records (5) belongs to customer 42 in Luma, 7 belongs to customer 42 in GQL implementation.
> 
> TODO: So when creating a new list for a customer, check for existing records belonging to the customer, if found (can be more than one), update those records with the newly created list_id. On above example, 5 should have list_id, 7 should have customer_id.
> 
> **Usecase 2:**
> When adding a product to the list in GQL, you are currently checking if the product already exists on the new implementation.
> 
> TODO: Nothing needs to be done for Luma change! if TODO on usecase 1 is performed this will work seamlessly.
> 
> **Usecase 3**
> This is where it gets tricky. When **all** the list items for the customer are deleted in Luma, the entry in compare_list table needs to be removed(just a single row, can exist, still). Some logic needs to be added for this.
> <img alt="Screen Shot 2020-11-02 at 5 21 30 PM" width="415" src="https://user-images.githubusercontent.com/43612864/97929915-e229f100-1d2f-11eb-887d-b1a817b3dc02.png">
> 
> If the above use cases are handled we can have an integrated system between Luma and the new GQL version.
> 
> @rogyar @cpartica, Please add your thoughts.

Hi @prabhuram93 
I have applied mentioned changes
Thanks

Usik2203 commented:
@magento run all tests

Usik2203 committed:
minor changes

Usik2203 commented:
@magento run all tests

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type Query {
+    compareList(uid: ID!): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CompareList") @doc(description: "Compare list")
```
```suggestion
    compareList(uid: ID!): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CompareList") @doc(description: "Return products that have been added to the specified compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+type Mutation {
+    createCompareList(input: CreateCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CreateCompareList") @doc(description: "Creates a new compare list. For a logged in user, the created list is assigned to the user")
+    addProductsToCompareList(input: AddProductsToCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddProductsToCompareList") @doc(description: "Add products to compare list")
```
```suggestion
    addProductsToCompareList(input: AddProductsToCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddProductsToCompareList") @doc(description: "Add products to the specified compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+type Mutation {
+    createCompareList(input: CreateCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CreateCompareList") @doc(description: "Creates a new compare list. For a logged in user, the created list is assigned to the user")
+    addProductsToCompareList(input: AddProductsToCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddProductsToCompareList") @doc(description: "Add products to compare list")
+    removeProductsFromCompareList(input: RemoveProductsFromCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\RemoveProductsFromCompareList") @doc(description: "Remove products from compare list")
```
```suggestion
    removeProductsFromCompareList(input: RemoveProductsFromCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\RemoveProductsFromCompareList") @doc(description: "Remove products from the specified compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+    createCompareList(input: CreateCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CreateCompareList") @doc(description: "Creates a new compare list. For a logged in user, the created list is assigned to the user")
+    addProductsToCompareList(input: AddProductsToCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddProductsToCompareList") @doc(description: "Add products to compare list")
+    removeProductsFromCompareList(input: RemoveProductsFromCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\RemoveProductsFromCompareList") @doc(description: "Remove products from compare list")
+    assignCompareListToCustomer(uid: ID!): Boolean @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AssignCompareListToCustomer") @doc(description: "Assign compare list to customer")
```
```suggestion
    assignCompareListToCustomer(uid: ID!): Boolean @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AssignCompareListToCustomer") @doc(description: "Assign the specified compare list to the logged in customer")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+    addProductsToCompareList(input: AddProductsToCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AddProductsToCompareList") @doc(description: "Add products to compare list")
+    removeProductsFromCompareList(input: RemoveProductsFromCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\RemoveProductsFromCompareList") @doc(description: "Remove products from compare list")
+    assignCompareListToCustomer(uid: ID!): Boolean @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\AssignCompareListToCustomer") @doc(description: "Assign compare list to customer")
+    deleteCompareList(uid: ID!): DeleteCompareListOutput @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\DeleteCompareList") @doc(description: "Delete compare list")
```
```suggestion
    deleteCompareList(uid: ID!): DeleteCompareListOutput @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\DeleteCompareList") @doc(description: "Delete the specified compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+input CreateCompareListInput {
+    products: [ID!]
```
```suggestion
    products: [ID!] @doc(description: "An array of product IDs to add to the compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+input AddProductsToCompareListInput {
+    uid: ID!,
+    products: [ID!]!
```
```suggestion
    products: [ID!]! @doc(description: "An array of product IDs to add to the compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+input RemoveProductsFromCompareListInput {
+    uid: ID!,
+    products: [ID!]!
```
```suggestion
    products: [ID!]! @doc(description: "An array of product IDs to remove from the compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+input AddProductsToCompareListInput {
+    uid: ID!,
```
```suggestion
    uid: ID!, @doc(description: "The unique identifier of the compare list to modify")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+input RemoveProductsFromCompareListInput {
+    uid: ID!,
```
```suggestion
    uid: ID!, @doc(description: "The unique identifier of the compare list to modify")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type DeleteCompareListOutput {
+    result: Boolean!
```
```suggestion
    result: Boolean! @doc(description: "Indicates whether the compare list was successfully assigned to the customer")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+# See COPYING.txt for license details.
+
+type ComparableItem {
+    uid: ID!
```
```suggestion
    uid: ID! @doc(description: "The unique ID of an item in a compare list")
```

prabhuram93 replied:
@keharper Please take a look at the schema descriptions. Thanks!

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+type ComparableItem {
+    uid: ID!
+    product: ComparableProduct!
```
```suggestion
    product: ComparableProduct! @doc(description: "Contains details about a product in a compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+type ComparableItem {
+    uid: ID!
+    product: ComparableProduct!
+    attributes: [ProductAttribute]! @doc(description: "Product comparable attributes")
```
```suggestion
    attributes: [ProductAttribute]! @doc(description: "An array of product attributes that can be used to compare products")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type ProductAttribute {
+    code: String! @doc(description:"Attribute code")
```
```suggestion
    code: String! @doc(description: "The unique identifier for a product attribute code.")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+type ProductAttribute {
+    code: String! @doc(description:"Attribute code")
+    value: String! @doc(description:"Attribute display value")
```
```suggestion
    value: String! @doc(description:"The display value of the attribute")
```

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

prabhuram93 committed:
- Modified assigncomparelust output to match schema in architecture
- Added validations for list_id of one customer cannot be accessed by other customer
- Added item_count
- tests pending for above

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type ComparableAttribute {
+    code: String! @doc(description: "Attribute code")
```
```suggestion
    code: String! @doc(description: "An attribute code that is enabled for product comparisons")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+type ComparableAttribute {
+    code: String! @doc(description: "Attribute code")
+    label: String! @doc(description: "Attribute label")
```
```suggestion
    label: String! @doc(description: "The label of the attribute code")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type CompareList {
+    uid: ID! @doc(description: "Compare list id")
```
```suggestion
    uid: ID! @doc(description: "The unique ID assigned to the compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+
+type CompareList {
+    uid: ID! @doc(description: "Compare list id")
+    items: [ComparableItem] @doc(description: "Comparable products")
```
```suggestion
    items: [ComparableItem] @doc(description: "An array of products to compare")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+type CompareList {
+    uid: ID! @doc(description: "Compare list id")
+    items: [ComparableItem] @doc(description: "Comparable products")
+    attributes: [ComparableAttribute] @doc(description: "Comparable attributes, provides codes and titles for the attributes")
```
```suggestion
    attributes: [ComparableAttribute] @doc(description: "An array of attributes that can be used for comparing products")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type Customer {
+    compare_list: CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CustomerCompareList") @doc(description: "Active customers compare list")
```
```suggestion
    compare_list: CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CustomerCompareList") @doc(description: "The contents of the customer's compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,61 @@
+}
+
+type Mutation {
+    createCompareList(input: CreateCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CreateCompareList") @doc(description: "Creates a new compare list. For a logged in user, the created list is assigned to the user")
```
```suggestion
    createCompareList(input: CreateCompareListInput): CompareList @resolver(class: "\\Magento\\CompareListGraphQl\\Model\\Resolver\\CreateCompareList") @doc(description: "Creates a new compare list. The compare list is saved for logged in customers")
```

keharper reviewed (commented):
All my suggestions are optional, but please accept them unless they are incorrect or if you can improve them. I recommend using the GH web page to accept suggestions in a batch, then merging that batch.

prabhuram93 committed:
- Updated assigncomparelist impl
- On assigning a compare list "A" to customer, if there already exist a list "B" for the customer, then all the items in "A" should be added to "B" and list "A" should be deleted

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

prabhuram93 committed:
Merge commit 'b21be308c2a369bcbcde2db5921faaf8b2b237af' into compare-products

prabhuram93 committed:
Merge commit '27a592cf38b501bcf05d9ef2e5401a9cff7e986b' into compare-products

Usik2203 committed:
Update schema description

Usik2203 committed:
minor changes

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

andrewbess commented on app/code/Magento/Catalog/etc/db_schema_whitelist.json:
```diff
@@ -1122,5 +1124,20 @@
+        }
     }
-}
+}
```
Hello @Usik2203 
Thank you for your contribution

Please don't forget to add a new line at end of file
```suggestion
}

```

Usik2203 replied:
This file was generated by Magento . 
It seems bad idea add empty line  at end of the file.
Thanks

andrewbess commented on app/code/Magento/CompareListGraphQl/composer.json:
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "magento/module-compare-list-graph-ql",
+    "version": "1.0.0",
```
Hello @Usik2203 
Thank you for your contribution

Please remove unneeded element `version`
I think it can help you resolving the problem with static tests.
```suggestion
```

Usik2203 replied:
**Done**

Usik2203 commented:
@magento run Integration Tests , Unit Tests tests

Usik2203 commented:
@magento run Unit Tests tests

Usik2203 commented:
@magento run Unit Tests

andrewbess commented on app/code/Magento/CompareListGraphQl/etc/module.xml:
```diff
@@ -0,0 +1,17 @@
+        <sequence>
+            <module name="Magento_Catalog"/>
+            <module name="Magento_Customer"/>
+            <module name="Magento_GraphQl"/>
```
Hello @Usik2203 
Thank you for your contribution.

Please add the sequence on Magento_CatalogGraphQl
```suggestion
            <module name="Magento_GraphQl"/>
            <module name="Magento_CatalogGraphQl"/>
```
Thank you in advance

Usik2203 replied:
Why we should add this dependancy ? I don't use any `Magento_CatalogGraphQl`  logic  in `Magento_CompareListGraphQl` .
 It seems be redundant.
Thanks

Usik2203 committed:
changes in composer.json

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 committed:
Update app/code/Magento/CompareListGraphQl/etc/module.xml

Co-authored-by: Andrii Beziazychnyi <a.beziazychnyi@atwix.com>

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

prabhuram93 committed:
compare-products - Fixed errors on creating comparelist and adding products.

dthampy committed:
MC-37399: Product Compare :: Atwix PR review and delivery

- add test coverage

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,64 @@
+    uid: ID! @doc(description: "The unique ID assigned to the compare list")
+    items: [ComparableItem] @doc(description: "An array of products to compare")
+    attributes: [ComparableAttribute] @doc(description: "An array of attributes that can be used for comparing products")
+    item_count: Int!
```
```suggestion
    item_count: Int! @doc(description: "The number of items in the compare list")
```

keharper commented on app/code/Magento/CompareListGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,64 @@
+
+type AssignCompareListToCustomerOutput {
+    result: Boolean!
+    compare_list: CompareList
```
```suggestion
    compare_list: CompareList @doc(description: "The contents of the customer's compare list")
```

keharper reviewed (commented):
A few more description updates

Usik2203 committed:
Update app/code/Magento/CompareListGraphQl/etc/schema.graphqls

Co-authored-by: Kevin Harper <keharper@users.noreply.github.com>

Usik2203 committed:
Update app/code/Magento/CompareListGraphQl/etc/schema.graphqls

Co-authored-by: Kevin Harper <keharper@users.noreply.github.com>

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

Usik2203 commented:
Hi @keharper 
All your recommendations were applied.
Thanks

prabhuram93 committed:
Merge remote-tracking branch 'usik/compare-products' into compare-products

prabhuram93 commented:
@magento run all tests

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

prabhuram93 committed:
Merge remote-tracking branch 'usik/compare-products' into compare-products

prabhuram93 commented:
@magento run all tests

prabhuram93 committed:
Merge remote-tracking branch 'mainline/2.4-develop' into compare-products

prabhuram93 commented:
@magento run all tests

prabhuram93 committed:
- Updated constraints on the catalog_compare_list table

prabhuram93 commented:
@magento run all tests

prabhuram93 committed:
Removing additional calls to the db.

Usik2203 committed:
Merge branch '2.4-develop' into compare-products

Usik2203 commented:
@magento run all tests

prabhuram93 commented:
@Usik2203 The changes are merged from out internal PR, but the PR is not because your last 2.4-develop merge commit is is not referenced in our internal PR. We will make sure to merge that commit on our next PR to 2.4-develop and this PR status will be updated. cc: @cpartica
